programming_assignments/api.py

```diff
@@ -0,0 +1,433 @@
+                "course"
+            ).get(id=assignment_id)
+        except SimpleProgrammingAssignment.DoesNotExist as e:
+            logger.exception(e)
+            return Response(str(e), status.HTTP_404_NOT_FOUND)
+        course_id = assignment.course_id
+
+        return self.create(request, course_id)
+
+    @action(detail=True, methods=["GET"])
+    def list_sections(self, request, pk):
+        """Gets all the section in the current assignment.
+
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): SimpleProgrammingAssignment id
+
+        Returns:
+            `Response` with all the section data and status `HTTP_200_OK`
+
+        Raises:
+            `HTTP_401_UNAUTHORIZED`: Raised by `IsInstructorOrTA` permission class
+            `HTTP_403_FORBIDDEN`: Raised due to `list()` method
+            `HTTP_404_NOT_FOUND`: Raised due to `list()` method
+        """
+        user = request.user
+        try:
+            assignment = SimpleProgrammingAssignment.objects.get(id=pk)
+        except SimpleProgrammingAssignment.DoesNotExist as e:
+            logger.exception(e)
+            return Response(str(e), status.HTTP_404_NOT_FOUND)
+
+        course_id = assignment.course_id
+        role = CourseHistory.objects.get(
+            course_id=course_id, user=user, status="E"
+        ).role
+        sections = []
+        if role == "S":
+            if assignment.is_published == "True":
+                sections = AssignmentSection.objects.filter(
+                    assignment=pk, section_type="V"
+                )
+        else:
+            sections = AssignmentSection.objects.filter(assignment=pk)
+        page = self.paginate_queryset(sections)
+        if page is not None:
+            serializer = self.get_serializer(page, many=True)
+            return self.get_paginated_response(serializer.data)
+        serializer = self.get_serializer(sections, many=True)
+        return Response(serializer.data)
+
+    @action(detail=True, methods=["GET"])
+    def retrieve_section(self, request, pk):
+        """Gets the section with primary key as pk to the instructor.
+           Gets the section with primary key as pk, only if it is visible
+           and assignment is published to the student.
+
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the section
+
+        Returns:
+            `Response` with the assignment data and status `HTTP_200_OK`
+
+        Raises:
+            `HTTP_401_UNAUTHORIZED`: Raised by:
+                1.`IsInstructorOrTA` permission class
+            `HTTP_403_FORBIDDEN`:  Raised:
+                1. If student is trying to access an unpublished assignment
+            `HTTP_404_NOT_FOUND`: Raised by `get_object()` method
+        """
+        user = request.user
+        section = self.get_object()
+        if (
+            CourseHistory.objects.get(user=user, course=section.assignment.course).role
```

Comment:
the above situation of list is handled here because it calls has_object_permisison